File: src/main/java/com/leet/easy/SameTree.java
package com.leet.easy;

/**
 * 100. Same Tree
 * https://leetcode.com/problems/same-tree/
 * Easy
 * Given the roots of two binary trees p and q, write a function to check if they are the same or not.
 * Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
 * Example 1:
 * Input: p = [1,2,3], q = [1,2,3]
 * Output: true
 * Example 2:
 * Input: p = [1,2], q = [1,null,2]
 * Output: false
 * Example 3:
 * Input: p = [1,2,1], q = [1,1,2]
 * Output: false
 * Constraints:
 *     The number of nodes in both trees is in the range [0, 100].
 *     -104 <= Node.val <= 104
 */

import com.leet.common.TreeNode;

public class SameTree {
  public boolean isSameTree(TreeNode p, TreeNode q) {
    if (p == null && q == null) return true;
    if (p == null || q == null) return false;
    return p.val == q.val && isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
  }
}

File: src/main/java/com/leet/easy/SameTree.exs
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec is_same_tree(p :: TreeNode.t() | nil, q :: TreeNode.t() | nil) :: boolean
  def is_same_tree(p, q) do
    cond do
      p == nil && q == nil -> true
      p == nil || q == nil -> false
      p.val != q.val -> false
      true -> is_same_tree(p.left, q.left) && is_same_tree(p.right, q.right)
    end
  end
end
